### PR TITLE
api/v1: Add drop_reason_desc enum to Flow API

### PR DESCRIPTION
**Pull request by kaworu**
Before this patch, `drop_reason` as a `uint32` was difficult to parse for humans.

This patch deprecate `drop_reason` and introduce `drop_reason_desc` as an enum value.

### CODE REVIEW

File: pkg/hubble/parser/seven/parser.go [hunk 1]
```diff
@@ -129,6 +129,7 @@ func (p *Parser) Decode(r *accesslog.LogRecord, decoded *pb.Flow) error {
 	decoded.Time = pbTimestamp
 	decoded.Verdict = decodeVerdict(r.Verdict)
 	decoded.DropReason = 0
+	decoded.DropReasonDesc = pb.DropReason_SUCCESS
```
Comment (glibsm) [inline pkg/hubble/parser/seven/parser.go]: `drop reason success` is rather confusing. Zero value should be left as some sort of an `UNKNOWN` if it's not applicable to the current type.

Besides, you don't have to set it explicitly.

File: pkg/monitor/api/drop.go [hunk 2]
```diff
@@ -18,7 +18,7 @@ import (
 	"fmt"
 )
 
-// DropMin numbers below this are non-drop reason codes
+// DropMin numbers lesser than this are non-drop reason codes
```
Comment (glibsm) [inline pkg/monitor/api/drop.go]: `less/smaller than`, but `below` was also fine

File: api/v1/flow/flow.proto [hunk 3]
```diff
@@ -310,22 +378,6 @@ enum EventType {
     RecordLost = 2;
 }
 
-// Payload is the payload, i.e. the data the monitor returns.
-message Payload {
```
Comment (glibsm) [inline api/v1/flow/flow.proto:320]: good catch

File: pkg/hubble/parser/seven/parser.go (hunk 1 above)
Comment (michi-covalent) [reply pkg/hubble/parser/seven/parser.go]: i guess this is trying to be consistent with https://github.com/cilium/cilium/blob/b3adc4d9410c76db934df021f99acaa033adf50f/pkg/monitor/api/drop.go#L27-L86

kind of curious how it shows up with `hubble observe -o json` though. will it show up like this?
```
"drop_reason_desc": "SUCCESS"
```
that would be a bit strange.

**Review by michi-covalent — APPROVED** [review]
lgtm just one minor question about the default values.

File: pkg/hubble/parser/seven/parser.go (hunk 1 above)
Comment (kaworu) [reply pkg/hubble/parser/seven/parser.go]: Maybe we should omit non-drop reasons from the enum, that way `drop_reason_desc` will be a missing field in theses cases. Would that make sense?

Comment (gandro) [reply pkg/hubble/parser/seven/parser.go]: We have to be careful with the zero (default) value. I'd prefer if we used `DROP_REASON_UNKNOWN = 0` to future proof it. According to this [post](http://androiddevblog.com/protocol-buffers-pitfall-adding-enum-values/), we ever add more drop reasons, client with older protobuf schemas will use the zero-valued enum to represent it. 

In that case, we want to avoid displaying `drop_reason: SUCCESS`. However, as you pointed out correctly, the zero value is treated as an missing value anyways, and not be displayed in the JSON output by default.

Edit: I just realized that my linked post refers to proto2. In proto3 it seems like the numeric value is preserved in Golang: https://developers.google.com/protocol-buffers/docs/proto3#enum - in that case I don't think it makes a big difference what name we used for the zero valued enum.

Comment (gandro) [reply pkg/hubble/parser/seven/parser.go]: Thinking about this some more, while it won't show up in JSON output by default, I think we should name it `DROP_REASON_UNKNOWN` to be consistent with all other enums that we have.

Calling it `SUCCESS` is a very leaky abstraction (it's originating in the kernel's errno `0`) and confusing to users of the API (they should not have to know or care about this). 

Calling it `DROP_REASON_UNKNOWN` also is I think the best way to stay backwards compatible. It also means that clients using a the new API talking to an old implementation will have drop events with non-zero drop reason, but no description. In these cases, we want the field to describe it as `DROP_REASON_UNKNOWN`, and not `SUCCESS` (which is wrong, as the flow was dropped).

Comment (michi-covalent) [reply pkg/hubble/parser/seven/parser.go]: +1 for `DROP_REASON_NONE`. `DROP_REASON_UNKNOWN` sounds a bit like the flow was dropped but the reason is unknown :)

Comment (kaworu) [reply pkg/hubble/parser/seven/parser.go]: > [...] In these cases, we want the field to describe it as DROP_REASON_UNKNOWN, and not SUCCESS (which is wrong, as the flow was dropped).

@gandro my understanding is that values [<130 are "non-drop reasons"](https://github.com/cilium/cilium/blob/master/pkg/monitor/api/drop.go#L21,L22) and so I don't think they can show up when the flow was dropped.

See also https://github.com/cilium/cilium/commit/8933b2f97cbff0e89998d1d045fac7974098993e where I would think that 2 is an exception (the packet is dropped).

Comment (gandro) [reply pkg/hubble/parser/seven/parser.go]: Okay, there are two independent issues here:

### Issue 1: The meaning of `non-drop` drop-reasons

There are two kinds of BPF perf events relevant to Hubble which have the numeric drop reason populated: `DropNotify` events, and `PolicyVerdict` events with a negative verdict ([source](https://github.com/cilium/cilium/blob/511b15d0e53cbe311465c624adbf9bc3fe492592/pkg/hubble/parser/threefour/parser.go#L343-L352))

The `DropMin` is used for the [`reason` field the metricsmap](https://github.com/cilium/cilium/blob/ada53a91a98618d24dad2abda97dd9c60923003c/bpf/lib/common.h#L248), which is overloaded to contain both drop reasons and non-drop-reasons. The metricsmap (not to be confused with the Hubble metrics) is irrelevant to Hubble at this point, as it never looks at the metricsmap.

The values that Hubble _can_ observe are the ones that the datapath produces for `DropNotify` and `PolicyVerdict` events, which are defined here:

https://github.com/cilium/cilium/blob/ada53a91a98618d24dad2abda97dd9c60923003c/bpf/lib/common.h#L327-L382

Therefore, I say we remove all the variants below 130 from the enum. The `errors` map in Go mixes up drop reasons and non-drop reasons simply for the [`DropForwardReason`](https://github.com/cilium/cilium/blob/fb0c3e592d7a9da854f9d38345fbb5720ea3b073/pkg/maps/metricsmap/metricsmap.go#L141) in the metricsmap. Again, nothing that is relevant to Hubble.

### Issue 2: The special meaning of a zero-valued enum in protobuf. 

 proto3 does not distinguish between an absent value and the zero-valued variant. This means that whenever you will use this new protobuf definition to talk to an older implementation of Hubble, *all* flows (dropped flows as well as forwarded l7 flows) will have `drop_reason_desc: SUCCESS` (even if it is not displayed in the JSON output because it's a zero-valued variant, it is still the value used in the Go API)

I don't think `drop_reason_desc: SUCCESS` makes sense for either forwarded or dropped flows. Let's quickly look at some the two cases:

1. Decoding a drop event from old Hubble implementation (which does not populate `drop_reason_desc`): `drop_reason` will be non-zero, but `drop_reason_desc` will be  `SUCCESS`. That's simply plain wrong. Here, `DROP_REASON_NONE` is also kinda wrong here (we do have a drop reason, but we don't know it's description), which is why I favor `DROP_REASON_UNKNOWN` over `DROP_REASON_NONE`.

2. Decoding a forwarded event (e.g. L7 or TraceNotify): `drop_reason_desc: SUCCESS` arguably makes some sense, but it's still confusing. What does it mean for a `drop` to be successful in a flow which was not dropped? Arguably, `DROP_REASON_NONE` would be the most correct here, `DROP_REASON_UNKNOWN` is at least not wrong.

### TL;DR: I propose the following changes:

1. Remove all non-drop reason variants from the enum (values below 130).
2. Let's name the zero-valued variant something else than `SUCCESS`. Either `DROP_REASON_UNKNOWN` (preferred by me) or `DROP_REASON_NONE`.

**Review by gandro — CHANGES_REQUESTED** [review]
https://github.com/cilium/cilium/pull/13301#discussion_r500853769

**Review by gandro — APPROVED** [review]

**Comment by kAworu** [comment]
test-me-please

**Review by michi-covalent — APPROVED** [review]
💯 let's go 🚀

**Comment by kAworu** [comment]
test-me-please

**Review by rolinh — APPROVED** [review]

**Comment by Rolinh** [comment]
retest-4.19

**Comment by aanm** [comment]
hit https://github.com/cilium/cilium/issues/13275
